### PR TITLE
[heft] Prototype of a method for plugins to hook into other plugins.

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: build-tests/heft-example-plugin-01/.heft/typescript.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/**
```
Comment (octogonz) [inline build-tests/heft-example-plugin-01/.heft/typescript.json:1]: Why did you add these config files that are entirely commented out?  

For Rush, we add blank templates of everything.  But for Heft, I think we want to minimize that, since otherwise people will copy+paste it as boilerplate to every project in the monorepo.

File: build-tests/heft-example-plugin-01/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tap(PLUGIN_NAME, () => {
+          this.hooks.exampleHook.call();
+        });
+      });
+    });
+  }
+}
+
+export default new ExamplePlugin01();
```
Comment (octogonz) [inline build-tests/heft-example-plugin-01/src/index.ts:36]: `export default` is deprecated because it does not enforce the name of the exported thing.

Might be better for the contract to use a name like:
```ts
export const heftPlugin = new ExamplePlugin01();
```

For API Documenter plugins we export a more rich manifest contract [like this](https://github.com/microsoft/rushstack/blob/5627415032539c27e93fdddac4edf93d2c87a311/repo-scripts/doc-plugin-rush-stack/src/index.ts#L7):
```ts
export const apiDocumenterPluginManifest: IApiDocumenterPluginManifest = {
  manifestVersion: 1000,
  features: [
    {
      featureName: 'rush-stack-markdown-documenter',
      kind: 'MarkdownDocumenterFeature',
      subclass: RushStackFeature
    }
  ]
};
```

Comment (octogonz) [reply build-tests/heft-example-plugin-01/src/index.ts:36]: BTW I'm planning to enable a lint rule to ban `export default` because I keep finding code that is using it.

File: build-tests/heft-example-plugin-02/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { IHeftPlugin, HeftSession, HeftConfiguration } from '@rushstack/heft';
+import { default as heftExamplePlugin01 } from 'heft-example-plugin-01';
+
+const PLUGIN_NAME: string = 'example-plugin-02';
```
Comment (octogonz) [inline build-tests/heft-example-plugin-02/src/index.ts:4]: If we make `IHeftPlugin` into proper class, then `PLUGIN_NAME` could be a static member, and we could eliminate this constant.  Instead we could use `ExamplePlugin02.displayName`.

Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:4]: Another idea would be to make `.tap` into a member function.  Then this:
```ts
     plugin.hooks.exampleHook.tap(PLUGIN_NAME, () => {
```
...becomes this:
```ts
     this.tap(plugin.hooks.exampleHook, () => {
```

Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:4]: That'd change the plugin import model, but that might be okay.

You still need to provide a name for each tap, though. So using this as a constant makes that easy.

File: apps/heft/src/pluginFramework/HeftSession.ts [hunk 4]
```diff
@@ -20,19 +22,19 @@ export interface IHeftSessionHooks {
...
-  metricsCollector: MetricsCollector;
-  getIsDebugMode(): boolean;
+  plugin: IHeftPlugin;
+  applyForPluginFn: ApplyForPluginFn;
 }
 
+/**
+ * @beta
+ */
+export type ApplyForPluginFn = <TPlugin extends IHeftPlugin>(
```
Comment (octogonz) [inline apps/heft/src/pluginFramework/HeftSession.ts:33]: `Fn` is not an English word.  Maybe call it `ApplyForPluginCallback`?

File: build-tests/heft-example-plugin-02/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { IHeftPlugin, HeftSession, HeftConfiguration } from '@rushstack/heft';
+import { default as heftExamplePlugin01 } from 'heft-example-plugin-01';
```
Comment (octogonz) [inline build-tests/heft-example-plugin-02/src/index.ts:2]: `default` is deprecated.  

Suppose the person wrote `import TypeScriptPlugin from 'heft-example-plugin-01';` thinking that they were importing a different plugin.  The compiler errors would be very confusing, because they would refer to it as `TypeScriptPlugin` instead of `ExamplePlugin01`.  Using name-based imports solves this problem, which is why `export default` should be deprecated.

As a second example, suppose you refactored by renaming `ExamplePlugin01` to be a different name. There is no way to automatically fix up all the existing imports -- they will compile without errors, continuing to use the old (wrong) name.

File: build-tests/heft-example-plugin-02/src/index.ts (hunk 3 above)
Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:4]: We'd have to change the plugin base to an abstract class, and this would deviate significantly from the way webpack plugins work. Both of those make me a little uncomfortable.

File: build-tests/heft-example-plugin-02/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+const PLUGIN_NAME: string = 'example-plugin-02';
+
+export class ExamplePlugin02 implements IHeftPlugin {
+  public displayName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    heftSession.applyForPlugin(heftExamplePlugin01, (plugin: typeof heftExamplePlugin01) => {
```
Comment (octogonz) [inline build-tests/heft-example-plugin-02/src/index.ts:10]: What happens if `heftExamplePlugin01` never actually gets loaded?  Is there any way for `ExamplePlugin02` to determine that?  Or would it simply wait indefinitely in hopes that it will eventually get loaded?

Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:10]: `heftExamplePlugin01` is an object reference, so this problem gets even more interesting in the case of side-by-side versions, where the plugin does get loaded, but with a different `prototype`.

File: build-tests/heft-example-plugin-02/src/index.ts (hunk 5 above)
Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:2]: `default` is the standard way to say "this is a special export." If you're exporting the plugin's type and the plugin instance, we need a way to say the instance is the important export.

File: build-tests/heft-example-plugin-02/src/index.ts (hunk 6 above)
Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:10]: The callback just won't get called. That's a scenario I'd expect the plugin developer(s) to have to handle themselves.

Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:10]: The side-by-side version issue is something I thought about, but I don't have a good solution for.

File: build-tests/heft-example-plugin-02/src/index.ts (hunk 5 above)
Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:2]: We came up with the idea that Heft plugin pacakges should include a "manifest" JSON file with metadata about each individual plugin and its entry point. This would enable Heft to inspect/enumerate the plugins exported by an NPM package without having to execute code.  The manifest can include a field the specifies the module path.

There's nothing special about the word `default`.  The special export could be like this:
```ts
export const heftPluginInstance:  IHeftPlugin = new new ExamplePlugin01();
```
...or like this:
```ts
export function createHeftPlugin(): IHeftPlugin {
  return new new ExamplePlugin01();
}
```

Or it could be a non-special name like this:

```ts
export class ExamplePlugin01 implements IHeftPlugin {
}
```
...and then the manifest would specify the export name:

**heft-plugin-manifest.json**
```ts
"plugins": [
  {
    "name": "Example Plugin",
    "modulePath": "./lib/ExamplePlugin.js",
    "exportName": "ExamplePlugin",
    "description": "..."
  }
]
```

Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:2]: (this design change should be a separate PR, however)

File: build-tests/heft-example-plugin-02/src/index.ts (hunk 6 above)
Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:10]: @iclanton What if we put some stakes in the ground like this:

- When a plugin hooks into another plugin, there should be a formal API contract that we try to keep stable.  This means:
- Plugins cannot be hooked into unless they explicitly agree to this in their manifest
- Plugins should explicitly separate their "public API" from the internal implementation details.
- Only Heft can run the `new ExamplePlugin02();` code; the consumer must request the public API object from Heft.  In a side-by-side scenario, this ensures the correct instances is returned.

File: build-tests/heft-example-plugin-02/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      plugin.hooks.exampleHook.tap(PLUGIN_NAME, () => {
+        heftConfiguration.globalTerminal.writeLine(
+          `!!!!!!!!!!!!!!! Plugin "${plugin.displayName}" hook called !!!!!!!!!!!!!!! `
```
Comment (octogonz) [inline build-tests/heft-example-plugin-02/src/index.ts:13]: @iclanton Here is a sketch of the above ideas:

**heft-example-plugin-01/src/index.ts**
```ts
export interface IExamplePlugin01Hooks {
  exampleHook: SyncHook;
}

/**
 * This defines the public contract for other plugins that want to access ExamplePlugin01.
 * Any changes to changes to this interface could break consumers.
 */
export interface IExamplePlugin01Accessor extends IHeftPluginAccessor {
  pluginId: string;
  hooks: IExamplePlugin01Hooks;
}

// This allows Heft to discover plugins without executing all their code.
export const examplePlugin01Manifest: IHeftPluginManifest = {
  pluginId: 'example-plugin-01',
  displayName: 'Example 1',
  entryPoint: './ExamplePlugin01.js'
};
```

**heft-example-plugin-01/src/ExamplePlugin01.ts**
```ts
// This object is private implementation details. Other plugins are NOT allowed to access it.
export class ExamplePlugin01 implements IHeftPlugin {
  public readonly displayName: string = examplePlugin01Manifest.displayName;

  public readonly hooks: IExamplePlugin01Hooks;

  constructor() {
    this.hooks = {
      exampleHook: new SyncHook()
    };
  }

  // Heft will call this at most once, if/when another plugin uses requestAccessToPlugin()
  // to access this plugin.  If `createAccessor` is not defined, then an error will be reported
  // saying that ExamplePlugin01 does not agree to be extended.
  public createAccessor(): IExamplePlugin01Accessor {
    return {
      manifest: examplePlugin01Manifest,
      hooks: this.hooks
    };
  }

  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
    heftSession.hooks.build.tap(examplePlugin01Manifest.pluginId, (build: IBuildStageContext) => {
      build.hooks.preCompile.tap(examplePlugin01Manifest.pluginId, (preCompile: IPreCompileSubstage) => {
        preCompile.hooks.run.tap(examplePlugin01Manifest.pluginId, () => {
          this.hooks.exampleHook.call();
        });
      });
    });
  }
}
```

**heft-example-plugin-02/src/index.ts**
```ts
// Note that we do NOT import the ExamplePlugin01 class.
// We only import the manifest metadata and the "public API"
import { examplePlugin01Manifest, IExamplePlugin01Accessor } from 'heft-example-plugin-01';

// This allows Heft to discover plugins without executing all their code.
export const examplePlugin02Manifest: IHeftPluginManifest = {
  pluginId: 'example-plugin-02',
  displayName: 'Example 2',
  entryPoint: './ExamplePlugin02.js'
};


export class ExamplePlugin02 implements IHeftPlugin {
  public displayName: string = examplePlugin02Manifest.pluginName;

  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
    // Here we ask Heft to locate the instance of "heft-example-plugin-01" and return its accessor.
    // If the plugin is not configured to be loaded, then Heft could report an error.
    heftSession.requestAccessToPlugin(examplePlugin01Manifest.pluginId, (examplePlugin01: IExamplePlugin01Accessor) => {
      examplePlugin01.hooks.exampleHook.tap(examplePlugin02Manifest.pluginId, () => {
        heftConfiguration.globalTerminal.writeLine(
          `!!!!!!!!!!!!!!! Plugin "${examplePlugin01.pluginId}" hook called !!!!!!!!!!!!!!! `
        );
      });
    });
  }
}
```

Comment (octogonz) [reply build-tests/heft-example-plugin-02/src/index.ts:13]: With the caveat that I still feel that calling `tap()` or `requestAccessToPlugin()` is a very important activity that Heft needs to oversee.  The caller should not need to constantly explain who they are.

For example, this code:
```ts
    heftSession.requestAccessToPlugin(
      PluginIds.ExamplePlugin02,  // who is attaching
      OtherPluginIds.ExamplePlugin01,  // whom we are attaching to
      (examplePlugin01: IExamplePlugin01Accessor) => {
      examplePlugin01.hooks.exampleHook.tap(PluginIds.ExamplePlugin02, () => {
```
...could be reduced to an API where the `this`=`PluginIds.ExamplePlugin02` so we don't need to specify our own `pluginId`:

```ts
    this.requestAccessToPlugin(OtherPluginIds.ExamplePlugin01,
      (examplePlugin01: IExamplePlugin01Accessor) => {
      this.attachHandler(examplePlugin01.hooks.exampleHook, () => {
```

Comment (iclanton) [reply build-tests/heft-example-plugin-02/src/index.ts:13]: It feels very awkward to enforce that plugins are classes extending from a base class because that means that plugins are married to a particular version of Heft and that plugins need to take a dependency on Heft. I'd rather keep the contract simple and let plugins be isolated with the theory that most plugins will perform simple/self-contained operations.
